docs/question/server.md

```diff
@@ -107,22 +339,35 @@ The `pl.to_json` function supports keyword-only options for different types of e
 From within `server.py` functions, directories can be accessed as:
 
 ```python
-data["options"]["question_path"]                      # on-disk location of the current question directory
-data["options"]["client_files_question_path"]         # on-disk location of clientFilesQuestion/
-data["options"]["client_files_question_url"]          # URL location of clientFilesQuestion/ (only in render() function)
-data["options"]["client_files_question_dynamic_url"]  # URL location of dynamically-generated question files (only in render() function)
-data["options"]["client_files_course_path"]           # on-disk location of clientFilesCourse/
-data["options"]["client_files_course_url"]            # URL location of clientFilesCourse/ (only in render() function)
-data["options"]["server_files_course_path"]           # on-disk location of serverFilesCourse/
+# on-disk location of the current question directory
+data["options"]["question_path"]
+
+# on-disk location of clientFilesQuestion/
+data["options"]["client_files_question_path"]
+
+# URL location of clientFilesQuestion/ (only in render() function)
+data["options"]["client_files_question_url"]
+
+# URL location of dynamically-generated question files (only in render() function)
+data["options"]["client_files_question_dynamic_url"]
+
+# on-disk location of clientFilesCourse/
+data["options"]["client_files_course_path"]
+
+# URL location of clientFilesCourse/ (only in render() function)
+data["options"]["client_files_course_url"]
+
+# on-disk location of serverFilesCourse/
+data["options"]["server_files_course_path"]
 ```
 
-## Generating dynamic files
+## Generating dynamic files with `file()`
 
-You can dynamically generate file objects in `server.py`. These files never appear physically on the disk. They are generated in `file()` and returned as strings, bytes-like objects, or file-like objects. A complete `question.html` and `server.py` example using a dynamically generated `fig.png` looks like:
+You can dynamically generate file objects in `server.py`. These files never appear physically on the disk. They are generated in `file()` and returned as strings, bytes-like objects, or file-like objects. `file()` has access to the full `data` object, including `data["params"]`. A complete `question.html` and `server.py` example using a dynamically generated `fig.png` looks like:
```

Comment:
`file()` does not have access to the whole object, AFAIK it does not have access to submitted answers, format errors, feedback, etc.

```suggestion
You can dynamically generate file objects in `server.py`. These files never appear physically on the disk. They are generated in `file()` and returned as strings, bytes-like objects, or file-like objects. `file()` has access to the same `data` object as the one set by the `generate()` function, including `data["params"]` and `data["correct_answers"]`. A complete `question.html` and `server.py` example using a dynamically generated `fig.png` looks like:
```

Reply:
Changed in 6408ede3ac9fe51ab8d193ea476887f9d181752a.